validate/validate.go

```diff
@@ -5,12 +5,103 @@ import (
 	"fmt"
 	"github.com/hashicorp/hcl/v2"
 	"io/ioutil"
+	"regexp"
 )
 
-// WordWithDotHyphenUnder80Pattern is a regexp pattern that matches string that contain alphanumerics, underscores, periods,
+type RegexpValidator struct {
+	pattern       string
+	errorTemplate string
+	regex         *regexp.Regexp
+}
+
+// Check validates provided string with a regexp based on the pattern and returns optional error.
+func (r *RegexpValidator) Check(value string, valueType interface{}) error {
+	if r.regex == nil {
+		var err error
+		r.regex, err = regexp.Compile(r.pattern)
+		if err != nil {
+			return err
+		}
+	}
+	if !r.regex.MatchString(value) {
+		return fmt.Errorf(r.errorTemplate, valueType)
+	}
+	return nil
+}
+
+// wordWithDotHyphenUnder80Pattern is a regexp pattern that matches string that contain alphanumerics, underscores, periods,
 // and hyphens that start with alphanumeric and End alphanumeric or underscore. Limits size to 1-80.
 // Based on https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules
-const WordWithDotHyphenUnder80Pattern = string(`^[a-zA-Z\d]$|^[a-zA-Z\d][\w\-.]{0,78}\w$`)
+const wordWithDotHyphenUnder80Pattern = string(`^[a-zA-Z\d]$|^[a-zA-Z\d][\w\-.]{0,78}\w$`)
+
+//NewWordWithDotHyphenUnder80Validator creates new RegexpValidator validating with wordWithDotHyphenUnder80Pattern.
+func NewWordWithDotHyphenUnder80Validator() *RegexpValidator {
+	return &RegexpValidator{wordWithDotHyphenUnder80Pattern, "%s can contain only alphanumerics, underscores, periods, and hyphens;" +
+		" must start with alphanumeric and end with alphanumeric or underscore and have 1-80 length", nil}
+}
+
+// cidrIPv4Pattern defines CIDR IPv4 notation with or without mask.
+const cidrIPv4Pattern = string(`^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)([\/][0-3][0-2]?|[\/][1-2][0-9]|[\/][0-9])?$`)
+
+//NewCIDRIPv4Check creates new RegexpValidator validating CIDR IPv4
+func NewCIDRIPv4Check() *RegexpValidator {
+	return &RegexpValidator{cidrIPv4Pattern, "%s not valid CIDR IPv4 value", nil}
+}
+
+// matchWholeWordsPattern creates OR words matching regexp pattern with words. Regexp special characters must be
+// escaped.
+func matchWholeWordsPattern(words []string) string {
+	var pattern string
+	for i, word := range words {
+		if len(word) == 0 {
+			continue
+		}
+		pattern += fmt.Sprintf(`^(%s)$`, word)
+		if i != len(words)-1 {
+			pattern += `|`
+		}
+	}
+	return pattern
+}
+
+func NewProtocolCheck() *RegexpValidator {
+	return &RegexpValidator{matchWholeWordsPattern([]string{"Tcp", "Udp", "Icmp", "Esp", "Ah", "\\*"}),
```

Comment:
Please change these to lowercase

Only include tcp, udp and icmp (and *) as the other ones are not supported by aws

Reply:
Should it be really be in lowercase? Followed https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/network_security_rule which states:

[protocol](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/network_security_rule#protocol) - (Required) Network protocol this rule applies to. Possible values include Tcp, Udp, Icmp, Esp, Ah or * (which matches all).

Reply:
Oh right, thats how you do it in Azure. But in Multy, we're following the AWS pattern of using lowercase instead. We can see that we transform that into the azure equivalente here: https://github.com/multycloud/multy/blob/main/resources/types/azure/network_security_group.go#L74